server/manager/src/main/java/org/apache/accumulo/manager/merge/FindMergeableRangeTask.java

```diff
@@ -0,0 +1,230 @@
+            if (!current.add(tm)) {
+              submit(current, type, table, namespaceId);
+              current.resetAndAdd(tm);
+            }
+          }
+
+          submit(current, type, table, namespaceId);
+        }
+
+      } catch (Exception e) {
+        log.error("Failed to generate system merges for {}", tableName, e);
+      }
+    }
+
+  }
+
+  void submit(MergeableRange range, FateInstanceType type, Entry<TableId,String> table,
+      NamespaceId namespaceId) {
+    if (range.tabletCount < 2) {
+      return;
+    }
+
+    log.debug("Table {} found {} tablets that can be merged for table", table.getValue(),
+        range.tabletCount);
+
+    TableId tableId = table.getKey();
+    String tableName = table.getValue();
+
+    range.startRow = Optional.ofNullable(range.startRow).map(Text::new).orElse(new Text(""));
+    range.endRow = Optional.ofNullable(range.endRow).map(Text::new).orElse(new Text(""));
+
+    String startRowStr = StringUtils.defaultIfBlank(range.startRow.toString(), "-inf");
+    String endRowStr = StringUtils.defaultIfBlank(range.endRow.toString(), "+inf");
+    log.debug("FindMergeableRangeTask: Creating merge op: {} from startRow: {} to endRow: {}",
+        tableId, startRowStr, endRowStr);
+    var fateId = manager.fate(type).startTransaction();
+    String goalMessage = TableOperation.MERGE + " Merge table " + tableName + "(" + tableId
+        + ") splits from " + startRowStr + " to " + endRowStr;
+
+    manager.fate(type).seedTransaction(FateOperation.SYSTEM_MERGE, fateId,
+        new TraceRepo<>(new TableRangeOp(Operation.SYSTEM_MERGE, namespaceId, tableId,
+            range.startRow, range.endRow)),
+        true, goalMessage);
+  }
+
+  static class MergeableRange {
+    final SteadyTime currentTime;
+    final long maxFileCount;
+    final long maxTotalSize;
+
+    Text startRow;
+    Text endRow;
+    int tabletCount;
+    long totalFileCount;
+    long totalFileSize;
+
+    MergeableRange(SteadyTime currentTime, long maxFileCount, long maxTotalSize) {
+      this.currentTime = currentTime;
+      this.maxFileCount = maxFileCount;
+      this.maxTotalSize = maxTotalSize;
+    }
+
+    boolean add(TabletMetadata tm) {
+      if (validate(tm)) {
+        tabletCount++;
+        log.trace("Adding tablet {} to MergeableRange", tm.getExtent());
+        if (tabletCount == 1) {
+          startRow = tm.getPrevEndRow();
+        }
+        endRow = tm.getEndRow();
+        totalFileCount += tm.getFiles().size();
+        totalFileSize += tm.getFileSize();
+        return true;
+      }
+      return false;
+    }
+
+    private boolean validate(TabletMetadata tm) {
+      if (tabletCount > 0) {
+        // If this is not the first tablet, then verify its prevEndRow matches
+        // the last endRow tracked, the server filter will skip tablets marked as never
+        if (!tm.getPrevEndRow().equals(endRow)) {
```

Comment:
`tm.getPrevEndRow()` could be null? `endRow` could be null?

Reply:
End row can definitely be null, prevEndRow _should_ not be null because the `if (tabletCount > 0) ` means it's at least the second tablet seen. However it would probably be good to add a null check just in case. If it is null we should throw an IllegalStateException probably.